File: python/AlgoExpert/semordnilap.py
def semordnilap(words):
    lookup = {}
    output = []
    for word in words:
        if word[::-1] not in lookup:
            lookup[word]=1
        else:
            del lookup[word[::-1]] 
            output.append([word[::-1],word])

    return output

# O(n*m) n is the length of words array and m is the length of the longest word
words = ["diaper", "abc", "test", "cba", "repaid"]
print(semordnilap(words))

File: javascript/AlgoExpert/semordnilap.js
// const semordnilap = (words) => {
//   const palindromes = {};

//   words.forEach((word) => {
//     const sortedWord = word.split("").sort().join("");
//     if (palindromes[sortedWord]) {
//       palindromes[sortedWord].push(word);
//     } else {
//       palindromes[sortedWord] = [word];
//     }
//   });

//   return Object.values(palindromes).filter((arr) => arr.length === 2);
// };

const semordnilap = (words) => {
  const wordSet = new Set(words);
  const result = [];
  words.forEach((word) => {
    const reversedWord = word.split("").reverse().join("");
    if (wordSet.has(reversedWord)) {
      result.push([word, reversedWord]);
      wordSet.delete(word);
      wordSet.delete(reversedWord);
    }
  });

  return result;
};

words = ["diaper", "abc", "test", "cba", "repaid"];
console.log(semordnilap(words));
